### PR TITLE
iter to stl

### PR DESCRIPTION
**Pull request by christopherlam**
~draft. too difficult~ to convert fully to `std::iterator` like methods.
EDIT: complete...

### CODE REVIEW

File: gnucash/import-export/import-main-matcher.cpp [hunk 1]
```diff
@@ -193,33 +195,83 @@ defer_bal_computation (GNCImportMainMatcher *info, Account* acc)
...
-    TreeViewIter (GtkTreeView *view)
-        : m_model{gtk_tree_view_get_model (view)}
-        , m_value{(bool)gtk_tree_model_get_iter_first (m_model, &m_iter)}
-    {};
-    gpointer get (gint col)
+    using iterator_category = std::forward_iterator_tag;
+    using difference_type = ptrdiff_t;
+    GtkTreeModelIterator(GtkTreeModel* model, std::optional<GtkTreeIter> iter)
+        : m_model(model), m_iter(iter) {}
+    GtkTreeModelIterator& operator++()
```
Comment (jralls) [inline gnucash/import-export/import-main-matcher.cpp]: Please add a blank line between functions!

**Comment by jralls** [comment]
Woot! A huge readability improvement!

File: gnucash/import-export/import-main-matcher.cpp [hunk 2]
```diff
@@ -192,32 +195,95 @@ defer_bal_computation (GNCImportMainMatcher *info, Account* acc)
     }
 }
 
+
+// Custom iterator class
+class GtkTreeModelIterator
+{
+public:
+    using iterator_category = std::forward_iterator_tag;
+    using difference_type = ptrdiff_t;
```
Comment (christopherlam) [inline gnucash/import-export/import-main-matcher.cpp]: these two lines are in the examples but removal doesn't harm.

File: gnucash/import-export/import-main-matcher.cpp [hunk 3]
```diff
@@ -192,32 +195,95 @@ defer_bal_computation (GNCImportMainMatcher *info, Account* acc)
...
+    {
+        if (m_iter.has_value() && !gtk_tree_model_iter_next(m_model, &*m_iter))
+            m_iter = std::nullopt;
+        return *this;
+    }
+    template <typename T>
+    T get (int column) {
+        g_return_val_if_fail (m_iter.has_value(), T());
+        gpointer rv;
+        gtk_tree_model_get(m_model, &*m_iter, column, &rv, -1);
```
Comment (christopherlam) [inline gnucash/import-export/import-main-matcher.cpp]: I'm using `std::nullopt` to denote the `end()` iterator. The original suggestions use `GtkTreeIter()` as the constructor for the `end()` iterator but I figure nullopt is safer, especially needing to test equality. This means we need the weird `&*m_iter` construct.

File: gnucash/import-export/import-main-matcher.cpp [hunk 4]
```diff
@@ -192,32 +195,95 @@ defer_bal_computation (GNCImportMainMatcher *info, Account* acc)
...
+// Custom container class
+class GtkTreeModelContainer
+{
+public:
+    using iterator = GtkTreeModelIterator;
+    GtkTreeModelContainer(GtkTreeModel* model) : m_model(model), m_size{0}
+    {
+        for (auto it = begin(); it.has_value(); ++it)
+            m_size++;
+    }
```
Comment (christopherlam) [inline gnucash/import-export/import-main-matcher.cpp]: this O(N) to calculate `size` is inevitable. Alternately we could forego size() and test empty via `return begin() == end()` in which both calls GtkTreeModelIterator constructor every time?

File: gnucash/import-export/import-main-matcher.cpp [hunk 5]
```diff
@@ -192,32 +195,95 @@ defer_bal_computation (GNCImportMainMatcher *info, Account* acc)
...
-            if (info->transaction_processed_cb)
-            {
-                info->transaction_processed_cb (trans_info, false,
-                                                info->user_data);
-            }
-        }
-        while (gtk_tree_model_iter_next (model, &iter));
-    }
+    if (info->transaction_processed_cb)
+        for (auto iter : GtkTreeModelContainer(gtk_tree_view_get_model (info->view)))
```
Comment (christopherlam) [inline gnucash/import-export/import-main-matcher.cpp]: instead of `for (auto iter : container)` I'd prefer `for (const auto& iter : container)` however the latter won't compile.

File: gnucash/import-export/import-main-matcher.cpp [hunk 6]
```diff
@@ -192,32 +195,106 @@ defer_bal_computation (GNCImportMainMatcher *info, Account* acc)
...
+    GtkTreeModel* m_model;
+    std::optional<GtkTreeIter> m_iter;
+};
+
+// Custom container class
+template <typename ModelType>
+class GtkTreeModelContainer
+{
+public:
+    using iterator = GtkTreeModelIterator<ModelType>;
```
Comment (jralls) [inline gnucash/import-export/import-main-matcher.cpp]: Capitalize type names.

File: gnucash/import-export/import-main-matcher.cpp (hunk 4 above)
Comment (jralls) [reply gnucash/import-export/import-main-matcher.cpp]: Constructing `end()` is trivial, so that would be much faster for `empty()`. This will anyway work to get the size only of a GtkListModel because a GtkTreeModel's iterators have children reflecting its tree structure.

File: gnucash/import-export/import-main-matcher.cpp (hunk 3 above)
Comment (jralls) [reply gnucash/import-export/import-main-matcher.cpp]: You could use `&m_iter.value()` instead.

File: gnucash/import-export/import-main-matcher.cpp (hunk 5 above)
Comment (jralls) [reply gnucash/import-export/import-main-matcher.cpp]: That's because the result of `GtkTreeModelContainer(gtk_tree_view_get_model (info->view))` is a temporary and you're not allowed to take references to temporaries. There's apparently a lifetime adjustment in C++23 that will allow it in for loops, see Temporary range expression in [this](https://en.cppreference.com/w/cpp/language/range-for). It will be a few years before we can use that, so a work around would be to create a local and iterate on that:
```
auto info_view{GtkTreeModelContainer(gtk_tree_view_get_model (info->view))};
for (const auto& iter : info_view)
  //…
```

Comment (christopherlam) [reply gnucash/import-export/import-main-matcher.cpp]: Doesn't work...
```
/home/chris/sources/gnucash/stable/gnucash/import-export/import-main-matcher.cpp:309:27: error: cannot bind non-const lvalue reference of type 'ImportMatcherType&' to an rvalue of type 'ImportMatcherType'
  309 |         for (auto& iter : tree_model)
      |                           ^~~~~~~~~~
```

File: gnucash/import-export/import-main-matcher.cpp (hunk 4 above)
Comment (christopherlam) [reply gnucash/import-export/import-main-matcher.cpp]: > This will anyway work to get the size only of a GtkListModel because a GtkTreeModel's iterators have children reflecting its tree structure.

Not quite sure what this means. Anyway, amended.

File: gnucash/gnome-utils/gnc-gtk-utils.hpp [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+
+    GtkTreeModelContainer(GtkTreeModel* model)
+        : m_model{model}
+        , m_begin_iter{TreeModelIterator(m_model, std::nullopt)}
+        , m_end_iter{TreeModelIterator(m_model, std::nullopt)}
+    {
+        GtkTreeIter iter;
+        gtk_tree_model_get_iter_first(m_model, &iter);
+        m_begin_iter = TreeModelIterator(m_model, iter);
+    }
```
Comment (christopherlam) [inline gnucash/gnome-utils/gnc-gtk-utils.hpp]: If this last commit is desirable, this ctor will init m_begin_iter with a dummy value before finding the proper beginning iterator. If we remove the initializer line then compilation fails.

**Comment by christopherlam** [comment]
I think the .hpp is now ready... I wouldn't merge the import-main-matcher.cpp until #1708

**Comment by jralls** [comment]
We need a more descriptive name for the header. `gnc-gtk-treeview-container.hpp` would be the normal choice because it's based on the name of the class it declares but it seems a bit wordy.

File: gnucash/import-export/import-main-matcher.cpp (hunk 5 above)
Comment (jralls) [reply gnucash/import-export/import-main-matcher.cpp]: Suggesting that you could bind a `const lvalue reference` but IIUC you already tried that. Looks like the compiler is optimizing away tree_model and leaving you with the temporary anyway. That probably means that `iter` is the temporary's memory and taking a reference wouldn't improve performance.

**Comment by gjanssens** [comment]
> We need a more descriptive name for the header. `gnc-gtk-treeview-container.hpp` would be the normal choice because it's based on the name of the class it declares but it seems a bit wordy.

`gnc-treeview-container.hpp` - whenever we inherit from gtk widgets we replace gtk with gnc. I think it will be ok to the the same here.

**Comment by christopherlam** [comment]
Ok. `gnc-tree-model-container.hpp` it is then. (not tree-view).

**Comment by jralls** [comment]
>  whenever we inherit from gtk widgets we replace gtk with gnc. I think it will be ok to the the same here.

It's composition not inheritance, but the `gtk` is reasonably implied. `view` and `model` are similarly irrelevant, so maybe just `gnc-tree-container.hpp` and the classes could be `GncTree(Container|Data|Iter)`.

**Comment by christopherlam** [comment]
> > whenever we inherit from gtk widgets we replace gtk with gnc. I think it will be ok to the the same here.
> 
> It's composition not inheritance, but the `gtk` is reasonably implied. `view` and `model` are similarly irrelevant, so maybe just `gnc-tree-container.hpp` and the classes could be `GncTree(Container|Data|Iter)`.

It's tailored for GtkTreeModel though, with `gtk_tree_model_get_iter_first` in it.

**Comment by jralls** [comment]
> It's tailored for GtkTreeModel though, with gtk_tree_model_get_iter_first in it.

Of course. What's your point?

**Comment by christopherlam** [comment]
> > It's tailored for GtkTreeModel though, with gtk_tree_model_get_iter_first in it.
> 
> Of course. What's your point?

It'd be appropriate to name it GncTreeModel?

**Comment by jralls** [comment]
> It'd be appropriate to name it GncTreeModel?

It would not, because it's not a stand-alone tree model. You can't create a GncTreeModel on its own, you have to create a GtkTreeModel and pass it in. It doesn't take ownership, and you need to keep the GtkTreeModel* handy so you can load it, mutate it, connect to its signals,  assign it to a GtkTreeView, and ultimately destroy it. You can't even safely keep it around because its `m_model` can be freed out from under it.

**Comment by christopherlam** [comment]
Ok `GncTreeContainer` it is then.

**Comment by christopherlam** [comment]
Re model it's probably better to pepper some `GTK_IS_TREE_MODEL` around then.

**Comment by jralls** [comment]
> Re model it's probably better to pepper some GTK_IS_TREE_MODEL around then.

Should be needed only in the constructor to make sure that you really have one. Since you have to write a ctor you might as well ref the thing and write a dtor to unref it or make it a unique_ptr with a custom deleter that unrefs it. That will protect it from being freed while the container is alive.
